### ISSUE TITLE
Crash as  assertion failed on iwyu 0.20

### ISSUE THREAD

**Issue by storypku** [include-what-you-use/include-what-you-use#1275]
Hi @kimgr ,
   Recently I came across an IWYU crash issue reporting:

```
iwyu: external/llvm-project/clang/lib/Basic/SourceManager.cpp:2027: bool clang::SourceManager::isBeforeInTranslationUnit(clang::SourceLocation, clang::SourceLocation) const: Assertion `LHS.isValid() && RHS.isValid() && "Passed invalid source location!"' failed.
```
If applied the following patch:

```
diff --git a/iwyu.cc b/iwyu.cc
index bbb6ac3..47b5630 100644
--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3234,11 +3234,19 @@ class InstantiatedTemplateVisitor
     // explicit template instantiations or already provided by the template.
     std::vector<const CXXRecordDecl*> explicit_inst_decls;
     for (const NamedDecl* redecl : decl->redecls()) {
-      if (IsExplicitInstantiation(redecl) &&
-          GlobalSourceManager()->isBeforeInTranslationUnit(
-            redecl->getLocation(), caller_loc())) {
-        // Earlier checks imply that this is a CXXRecordDecl.
-        explicit_inst_decls.push_back(cast<CXXRecordDecl>(redecl));
+      if (IsExplicitInstantiation(redecl)) {
+        const SourceLocation redecl_loc = redecl->getLocation();
+        const SourceLocation caller_loc_ = caller_loc();
+        if (redecl_loc.isInvalid() or caller_loc_.isInvalid()) {
+          VERRS(6) << "Skip invalid template location as either redec_loc or caller_loc is invalid:  "
+            << redecl->getNameAsString()  << "\n";
+          continue;
+        }
+
+        if (GlobalSourceManager()->isBeforeInTranslationUnit(redecl_loc, caller_loc_)) {
+          // Earlier checks imply that this is a CXXRecordDecl.
+          explicit_inst_decls.push_back(cast<CXXRecordDecl>(redecl));
+        }
       }
     }
 ```

The most relevant error log with above patch (`--verbose=6` for iwyu) is attached below:
[iwyu_source_location_invalid.txt](https://github.com/include-what-you-use/include-what-you-use/files/12095123/iwyu_source_location_invalid.txt)  (Notice the `Skip invalid template location` keywords).

Sorry I can't find a minimal reproducible case temporarily. My working environment is Ubuntu 18.04 x86_64, with Clang 16.0.0 and IWYU 0.20 release. GCC is 9.3.0 if relevant. If further context is needed, I would like to help.

Thanks in advance for investigating into it.

**Comment by kimgr** [comment]
Superficially the patch makes sense. I am away from source code right now, but I need to look into this a little closer. It's possible that the missing source loc is a deficiency in clang that we could consider improving.

**Comment by kimgr** [comment]
Just to confirm: you are not using precompiled headers, are you?

**Comment by storypku** [comment]
> 
No precompiled headers is used. I think in my situation, the reason was that I built   IWYU with LLVM sources in debug mode (i.e., NDEBUG undefined), so the assertion inside the impl of `isBeforeInTranslationUnit` show up.

**Comment by UniQP** [comment]
Here's a small test case:

`test.h`
```
#include <iostream>

enum E { A };
std::ostream& operator<<(std::ostream&, E);

#define macro std::cout << A
```

`test.cpp`
```
#include "test.h"

void test() {
  macro;
}
```

```
include-what-you-use --version
include-what-you-use 0.21 (git:0170be4) based on clang version 17.0.0 (https://github.com/llvm/llvm-project.git 3d2b45620ba97c77c15292965878a6f714c71a81)

include-what-you-use test.cpp
include-what-you-use: […]/llvm-project/clang/lib/Basic/SourceManager.cpp:2026: bool clang::SourceManager::isBeforeInTranslationUnit(clang::SourceLocation, clang::SourceLocation) const: Assertion `LHS.isValid() && RHS.isValid() && "Passed invalid source location!"' failed.
```

The problem happens with #1294 as well but not with the changes suggested by @storypku.

**Comment by kimgr** [comment]
@UniQP Thanks for the repro case! OK, so it seems the problem is not the decl location, but the actual `caller_loc()` which is the current location in the AST. It's like Clang gets lost in its own tree 🥲.

Will try to update the PR when I have some spare time.

**Comment by kimgr** [comment]
I had some time to debug this, and it looks like we hit this path: https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_ast_util.cc#L200. Since we're in a macro, IWYU doesn't know whether it should report the location for expansion or for spelling, and so gives up and returns an invalid location.

I think we'll go with something like @storypku's patch, the advisory logging can be used to diagnose when/if a wrong decision for explicit instantiation is made due to missing locations.

**Comment by kimgr** [comment]
Updated PR #1294 , please take another look and retest.

**Comment by kimgr** [comment]
Merged to master as of 46f908ff2ad1b6da226c8190624c03d76b7eee9a
